File: Graph/TwoEdgeConnectedComponents.py
from typing import List
 
class TwoEdgeConnectedComponents:
    def __init__(self, n:int, G:List[List[int]]):
        self.n = n
        self.seen = [False]*n
        self.check_aps = [False]*n
        self.ord = [0]*n
        self.low = [0]*n
        self.G = G
        self.groups = []
        self.idx = [0]*n
        k = 0
        for i in range(n):
            if not self.seen[i]:
                self.dfs(i,-1,k)
        self.build()
    
    def build(self) -> None:
        self.seen = [False]*self.n
        cnt = 0
        for i in range(self.n):
            if self.seen[i]:
                continue
            q = [i]
            self.groups.append([])
            for j in q:
                if self.seen[j]:
                    continue
                self.groups[cnt].append(j)
                self.idx[j] = cnt
                self.seen[j] = True
                for k in self.G[j]:
                    if self.seen[k]:
                        continue
                    if self.is_bridges(j,k):
                        continue
                    q.append(k)
            cnt += 1

    def get_idx(self, x:int) -> int:
        return self.idx[x]
 
    def dfs(self, now:int, par:int, k:int) -> int:
        self.seen[now] = True
        self.ord[now] = k
        self.low[now] = k
        k += 1
        count = 0
        flag = True
        for i in self.G[now]:
            if i == par and flag:
                flag = False
                continue
            if self.seen[i]:
                self.low[now] = min(self.low[now],self.ord[i])
            else:
                count += 1
                k = self.dfs(i,now,k)
                self.low[now] = min(self.low[now],self.low[i])
                if par != -1 and self.ord[now] <= self.low[i]:
                    self.check_aps[now] = True
        if par == -1 and count >= 2:
            self.check_aps[now] = True
        return k
 
    def is_aps(self, x:int) -> bool:
        return self.check_aps[x]
    
    def is_bridges(self, x:int, y:int) -> bool:
        if self.ord[x] > self.ord[y]:
            x,y = y,x
        return self.ord[x] < self.low[y]

File: Graph/TwoEdgeConnectedComponents.cpp
#include<bits/stdc++.h>
using namespace std;

struct TwoEdgeConnectedComponents{
  vector<vector<int>> groups;
  vector<int> idx;
 
  explicit TwoEdgeConnectedComponents(int _n,vector<vector<int>> _G): n(_n),G(_G) {
    seen.assign(n,false);
    check_aps.assign(n,false);
    ord.assign(n,0);
    low.assign(n,0);
    idx.assign(n,0);
    int k = 0;
    for(int i = 0;i < n;++i){
      if (!seen[i]) dfs(i,-1,k);
    }
    build();
  }

  void build(){
    seen.assign(n,false);
    int cnt = 0;
    queue<int> q;
    for(int i = 0;i < n;++i){
      if (seen[i]) continue;
      q.push(i);
      groups.push_back(vector<int>{});
      while(!q.empty()){
        int l = q.front();
        q.pop();
        if(seen[l]) continue;
        groups[cnt].push_back(l);
        idx[l] = cnt;
        seen[l] = true;
        for(int j:G[l]){
          if (seen[j]) continue;
          if (is_bridges(l,j)) continue;
          q.push(j);
        }
      }
      ++cnt;
    }
  }

  int get_idx(const int &k) {
    return idx[k];
  }
 
  void dfs(int now,int par,int &k){
    seen[now] = true;
    int count = 0;
    bool flag = true;
    ord[now] = k++;
    low[now] = ord[now];
    for(int i:G[now]){
      if (i == par && flag) {
        flag = false;
        continue;
      }
      if (seen[i]){
        low[now] = min(low[now],ord[i]);
      }else{
        ++count;
        dfs(i,now,k);
        low[now] = min(low[now],low[i]);
        if (par != 1 && ord[now] <= low[i]) check_aps[now] = true;
      }
    }
    if (par == 1 && count >= 2) check_aps[now] = true;
  }
 
  bool is_aps(int x){
    return check_aps[x];
  }

  bool is_bridges(int x, int y){
    if (ord[x]>ord[y]) swap(x,y);
    return ord[x] < low[y];
  }
 
  private:
    const int n;
    const vector<vector<int>> G;
    vector<int> ord,low;
    vector<bool> seen,check_aps;
};